### PR TITLE
Rebase of previous fixes

### PR DESCRIPTION
**Pull request by FraMecca**
As discussed by email, this rebase includes the fixes of the past month.

I want to be clear that tests and benchmarks run successfully but I would prefer to discuss them after merging with upstream.
The master branch has advanced quite a bit and some programs such as dustmite have changed parts of their code so newer versions don't even compile.

### CODE REVIEW

File: src/gc/bits.d [hunk 1]
```diff
@@ -31,24 +31,27 @@ struct GCBits
...
             data = null;
         }
     }
 
     void alloc(size_t nbits) nothrow
     {
+        import gc.os : os_mem_map;
         this.nbits = nbits;
-        data = cast(typeof(data[0])*)calloc(nwords, data[0].sizeof);
+        data = cast(typeof(data[0])*)os_mem_map(nwords * data[0].sizeof, true); // Allocate as MAP_SHARED
```
Comment (llucax) [inline src/gc/bits.d]: I think the memory should only be shared for mark bits, all the rest is better to keep them private to the process. In tangort there is an option to allocate the bits as shared or private memory.

File: src/gc/bits.d [hunk 2]
```diff
@@ -31,24 +31,27 @@ struct GCBits
     wordtype* data;
     size_t nbits;
 
-    void Dtor() nothrow
+    void Dtor() nothrow @nogc
```
Comment (llucax) [inline src/gc/bits.d]: Not a big deal, but these `@nogc` are actually unrelated to this commit, right?

File: src/gc/os.d [hunk 3]
```diff
@@ -146,15 +146,16 @@ else static if (is(typeof(mmap)))  // else version (GC_Use_Alloc_MMap)
 {
     enum { HAVE_FORK = true }
 
-    void *os_mem_map(size_t nbytes) nothrow
+    void *os_mem_map(size_t nbytes, bool share = false) nothrow
```
Comment (llucax) [inline src/gc/os.d]: I prefer the more explicit approach used in tangort using an enum, but this is just bikeshedding :)

File: src/gc/impl/conservative/gc.d [hunk 4]
```diff
@@ -2149,7 +2155,7 @@ struct Gcx
         {
             pool = pooltable[n];
             pool.mark.zero();
-            if(!pool.isLargeObject) pool.freebits.zero();
+            // if (!pool.isLargeObject) pool.freebits.zero();
```
Comment (llucax) [inline src/gc/impl/conservative/gc.d]: This is commented because now you keep them updated at all times, right?

File: src/gc/stats.d [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+ *     in a product, an acknowledgment in the product documentation would be
+ *     appreciated but is not required.
+ *  o  Altered source versions must be plainly marked as such, and must not
+ *     be misrepresented as being the original software.
+ *  o  This notice may not be removed or altered from any source
+ *     distribution.
+ * Authors:   Walter Bright, Sean Kelly, Leandro Lucarella
+ */
+
+module gc.stats;
```
Comment (llucax) [inline src/gc/stats.d]: There is some overlap between the current stats collected by the GC and this, so I wonder how upstream will take this. But it proved to be very helpful in tangort.

File: src/gc/impl/conservative/gc.d [hunk 6]
```diff
@@ -2557,6 +2559,8 @@ Lmark:
                 }
             }
 
+            if(shouldFork)
+                thread_suspendAll();
```
Comment (llucax) [inline src/gc/impl/conservative/gc.d]: Threads are not suspended if it's not forking?

Also, I don't get how this is related to `minimize()`.

**Review by llucax — COMMENTED** [review]
This seems much cleaner than the previous PR, but still I find the commit message too terse. Sometimes if very hard to review a commit when the intention of what's going on is not stated clearly in the commit message (not only title). For example I have no idea about what's going on in commit `more concurrent behaviour with respect to minimize()` with `shouldFork`.

File: src/gc/impl/conservative/gc.d (hunk 4 above)
Comment (FraMecca) [reply src/gc/impl/conservative/gc.d]: yep, this was actually suggested by you on email

File: src/gc/stats.d (hunk 5 above)
Comment (FraMecca) [reply src/gc/stats.d]: I think that upstream will refuse this but it is needed for our benchmarking and discussion.

**Comment by FraMecca** [comment]
Thank you Leandro.

If you prefer I can close the pull and rebase adding better commit messages.

If it is enough for you I'll explain the commits here:

 * [added stats.d to SRCS](https://github.com/FraMecca/druntime/pull/3/commits/023fec0b42d3d8f42a3707894a9d91fc9c56dd37): SRCS are used by the runtime to list the files to be compiled.

* [stats.d from tangort](https://github.com/FraMecca/druntime/pull/3/commits/e649d812ccd2e34caf1dc5784c3e685bf34871fb): taken from tangort, needs some functions to work properly and they are added in the other commits.

* [more concurrent behaviour with respect to minimize()](https://github.com/FraMecca/druntime/pull/3/commits/1a04181808d4a7f389b34f99bafb306c69b0c9d2): minimize() should be called after a fullcollect. For this reason I set a boolean shouldMinimize that is checked at the end of the fullcollect function.

* [options for concurrent behaviour](https://github.com/FraMecca/druntime/pull/3/commits/6bff6e3aadf420c8613381cc1b99cf1785103e21): we discussed this by email for many weeks. This should cover all the cases where allocations are being done and changes should be reflected on markbits and freebits.
 
* [reflect functions in gcx from main class](https://github.com/FraMecca/druntime/pull/3/commits/f685afa3c58efbd73b01913a2ca896f2f680bd35): the main GC class functions now have an option for forking that is defaulted to false.

* [options for concurrent behaviour](https://github.com/FraMecca/druntime/pull/3/commits/6bb22be594682e6650a1437fa7ccbb8a4a1d7a5d): forking behaviour can be enabled by a commandline switch.

* [allocate gcbits as mmap shared](https://github.com/FraMecca/druntime/pull/3/commits/89f8b4e16ed50d80b9cfa6e5c89d22a57453a8b7): as discussed by email, needed in order for the children to work correctly and notify its changes to the parent.

**Comment by llucax** [comment]
I think is a good idea to always write more comprehensive commit messages in general, it really eases the process of reviewing code. In particular if you want to merge this upstream later on, it won't be a wasted effort.

You don't need to close this PR, you can rebase and force-push.

Thanks!

**Comment by FraMecca** [comment]
Hi Leandro, 
I have branched, rebased and merged.

Looking forward to your review.

File: src/gc/bits.d [hunk 7]
```diff
@@ -27,28 +28,36 @@ struct GCBits
     enum BITS_SHIFT = (wordtype.sizeof == 8 ? 6 : 5);
     enum BITS_MASK = (BITS_PER_WORD - 1);
     enum BITS_1 = cast(wordtype)1;
+    bool mmap;
```
Comment (llucax) [inline src/gc/bits.d]: Why do you need this bool?

File: src/gc/bits.d [hunk 8]
```diff
@@ -27,28 +28,36 @@ struct GCBits
...
     size_t nbits;
 
-    void Dtor() nothrow
+    void Dtor() nothrow @nogc
     {
         if (data)
         {
-            free(data);
+            if (mmap)
+                os_mem_unmap(data, nwords * data[0].sizeof);
```
Comment (llucax) [inline src/gc/bits.d]: Ohhh. Why don't you just use mmap for everything instead? Is it to save memory because usually bitmaps are less than 4k?

File: src/gc/impl/conservative/gc.d [hunk 9]
```diff
@@ -1866,6 +1872,8 @@ struct Gcx
         if (pool)
         {
             pool.initialize(npages, isLargeObject);
+            if (collectInProgress)
+                pool.mark.setAll();
```
Comment (llucax) [inline src/gc/impl/conservative/gc.d]: Why was this moved here?

File: src/gc/impl/conservative/gc.d [hunk 10]
```diff
@@ -1131,12 +1131,24 @@ class ConservativeGC : GC
     }
 
 
+    void collect(bool block = true) nothrow
+    {
+        fullCollect(block);
+    }
+
+
     void collect() nothrow
```
Comment (llucax) [inline src/gc/impl/conservative/gc.d]: Why not just extending this one? Aren't both conflicting when `collect()` is called since the new function has a default value?

**Review by llucax — COMMENTED** [review]
Thanks, that helped to understand some of the changes! Besides the small comments LGTM in general.
